third-party/Makefile

```diff
@@ -88,31 +107,42 @@ else ifeq ($(wildcard $(GMP_H_FILE)),)
 endif
 
 gmp: $(GMP_H_FILE)
-$(GMP_H_FILE):
+$(GMP_H_FILE): $(GMP_DEPEND)
+	rm -rf $(GMP_INSTALL_DIR) && rm -rf $(GMP_BUILD_DIR)
 	cd gmp && $(MAKE)
 
 hwloc: $(HWLOC_INSTALL_DIR)
-$(HWLOC_INSTALL_DIR):
+$(HWLOC_INSTALL_DIR): $(HWLOC_DEPEND)
+	rm -rf $(HWLOC_INSTALL_DIR) && rm -rf $(HWLOC_BUILD_DIR)
 	cd hwloc && $(MAKE)
 
 jemalloc: $(JEMALLOC_INSTALL_DIR)
-$(JEMALLOC_INSTALL_DIR):
+$(JEMALLOC_INSTALL_DIR): $(JEMALLOC_DEPEND)
+	rm -rf $(JEMALLOC_INSTALL_DIR) && rm -rf $(JEMALLOC_BUILD_DIR)
 	cd jemalloc && $(MAKE)
 
 libunwind: $(LIBUNWIND_INSTALL_DIR)
-$(LIBUNWIND_INSTALL_DIR):
+$(LIBUNWIND_INSTALL_DIR): $(LIBUNWIND_DEPEND)
+	rm -rf $(LIBUNWIND_INSTALL_DIR) && rm -rf $(LIBUNWIND_BUILD_DIR)
 	cd libunwind && $(MAKE)
 
 massivethreads: $(MASSIVETHREADS_INSTALL_DIR)
-$(MASSIVETHREADS_INSTALL_DIR):
+$(MASSIVETHREADS_INSTALL_DIR): $(MASSIVETHREADS_DEPEND)
+	rm -rf $(MASSIVETHREADS_INSTALL_DIR) && rm -rf $(MASSIVETHREADS_BUILD_DIR)
 	cd massivethreads && $(MAKE)
 
+# Qthreads may use hwloc and/or jemalloc. Ensure they are built first
 ifeq ($(CHPL_MAKE_HWLOC), hwloc)
-qthread: $(HWLOC_INSTALL_DIR) $(QTHREAD_INSTALL_DIR)
-else
-qthread: $(QTHREAD_INSTALL_DIR)
+QTHREAD_DEPEND_INSTALL_DIRS += $(HWLOC_INSTALL_DIR)
 endif
-$(QTHREAD_INSTALL_DIR):
+ifeq ($(CHPL_MAKE_JEMALLOC), jemalloc)
+QTHREAD_DEPEND_INSTALL_DIRS += $(JEMALLOC_INSTALL_DIR)
+endif
+QTHREAD_DEPEND_INSTALL_DIRS += $(QTHREAD_INSTALL_DIR)
+qthread: $(QTHREAD_DEPEND_INSTALL_DIRS)
+
+$(QTHREAD_INSTALL_DIR): $(QTHREAD_DEPEND) $(HWLOC_DEPEND) $(JEMALLOC_DEPEND)
```

Comment:
I'll probably punt on this --

The expensive part is the actual `find` command, which we'll do regardless. Make seems to be pretty fast at computing dependencies. 

And if somebody is using qthreads, 99% of the time (my complete fabrication/guess) they'll be using hwloc and likely jemalloc. I imagine most users are either using the quickstart configuration or the default configuration, not some mix of qthreads and no hwloc.

What I thought you were suggesting was conditionalizing the find commands.. which make we could do now that the find_3p_deps is function-ified?